src/gwt/src/org/rstudio/core/client/command/UserCommandManager.java

```diff
@@ -0,0 +1,147 @@
+      
+      private final String name_;
+      private final Command command_;
+   }
+   
+   public static class UserCommandResult extends JavaScriptObject
+   {
+      protected UserCommandResult() {}
+      
+      public native final String getAction() /*-{ return this.action; }-*/;
+      public native final String getText() /*-{ return this.text; }-*/;
+      private native final JsArrayInteger getRangeVector() /*-{ return this.range; }-*/;
+      public final Range getRange()
+      {
+         JsArrayInteger rangeVector = getRangeVector();
+         return Range.create(
+               rangeVector.get(0),
+               rangeVector.get(1),
+               rangeVector.get(2),
+               rangeVector.get(3));
+      }
+   }
+   
+   public UserCommandManager()
+   {
+      RStudioGinjector.INSTANCE.injectMembers(this);
+      commandMap_ = new HashMap<KeyboardShortcut, UserCommand>();
+      
+      events_.addHandler(
+            RegisterUserCommandEvent.TYPE,
+            new RegisterUserCommandEvent.Handler()
+            {
+               @Override
+               public void onRegisterUserCommand(RegisterUserCommandEvent event)
+               {
+                  UserCommandManager.this.onRegisterUserCommand(event);
+               }
+            });
+   }
+   
+   @Inject
+   public void initialize(EventBus events,
+                          FilesServerOperations files)
+   {
+      events_ = events;
+      files_ = files;
+   }
+   
+   public boolean dispatch(KeyboardShortcut shortcut)
+   {
+      if (commandMap_.containsKey(shortcut))
+      {
+         UserCommand command = commandMap_.get(shortcut);
+         command.execute();
+         return true;
+      }
+      
+      return false;
+   }
+   
+   private void onRegisterUserCommand(RegisterUserCommandEvent event)
+   {
+      final String name = event.getData().getName();
+      JsArrayString shortcutStrings = event.getData().getShortcuts();
+      
+      for (int i = 0; i < shortcutStrings.length(); i++)
+      {
+         String shortcutString = shortcutStrings.get(i);
+         KeySequence sequence = KeySequence.fromShortcutString(shortcutString);
+         assert sequence != null : "Failed to parse string '" + shortcutString + "'";
+         
+         KeyboardShortcut shortcut = new KeyboardShortcut(sequence);
+         UserCommand command = new UserCommand(name, new Command()
+         {
+            @Override
+            public void execute()
+            {
+               events_.fireEvent(new ExecuteUserCommandEvent(name));
+            }
+         });
+         
+         commandMap_.put(shortcut, command);
+      }
+   }
+   
+   public Set<KeyboardShortcut> getKeyboardShortcuts()
+   {
+      return commandMap_.keySet();
+   }
+   
+   public Map<KeyboardShortcut, UserCommand> getCommands()
+   {
+      return commandMap_;
+   }
+   
+   public void saveBindings()
+   {
+      // TODO: Save binding list as well as the R environment
```

Comment:
Do we need these TODOs before merge?  I'll probably assume not and merge anyway but let's clean these up right away.